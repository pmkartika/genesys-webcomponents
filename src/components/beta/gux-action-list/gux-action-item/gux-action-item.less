@import (reference) '../../../../style/color.less';

:host {
  outline: none;
}

.gux-action-item {
  display: flex;
  width: 100%;
  height: max-content;
  padding: 0 16px;
  font-family: inherit;
  font-size: inherit;
  line-height: 32px;
  text-align: left;
  word-wrap: break-word;
  cursor: pointer;
  background-color: inherit;
  border: none;

  &:focus-within:not([disabled]),
  &:active:not([disabled]),
  &:hover:not([disabled]) {
    color: @gux-type-dark;
    background: @gux-blue-60;
    outline: none;
  }

  .gux-text {
    width: 100%;
  }

  &.gux-disabled {
    pointer-events: none;
    cursor: default;
  }
}
