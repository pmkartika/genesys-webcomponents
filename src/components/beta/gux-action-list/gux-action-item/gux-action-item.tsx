import {
  Component,
  Event,
  EventEmitter,
  h,
  Host,
  JSX,
  Listen,
  Prop
} from '@stencil/core';

@Component({
  styleUrl: 'gux-action-item.less',
  tag: 'gux-action-item',
  shadow: true
})
export class GuxActionItem {
  @Prop()
  text: string;

  @Prop()
  value: unknown;

  @Prop()
  disabled: boolean = false;

  @Event()
  press: EventEmitter<unknown>;

  @Listen('click')
  handleClick() {
    this.onItemClicked();
  }

  @Listen('keydown')
  onKeydown(event: KeyboardEvent): void {
    switch (event.key) {
      case 'Enter':
        event.preventDefault();
        this.onItemClicked();
        return;
    }
  }

  @Listen('keyup')
  onKeyup(event: KeyboardEvent): void {
    switch (event.key) {
      case ' ':
        event.preventDefault();
        this.onItemClicked();
        return;
    }
  }

  private onItemClicked(): void {
    if (!this.disabled) {
      this.press.emit(this.value);
    }
  }

  render(): JSX.Element {
    return (
      <Host role="listitem">
        <button
          disabled={this.disabled}
          onClick={() => this.onItemClicked()}
          class={{
            'gux-action-item': true,
            'gux-disabled': this.disabled
          }}
        >
          {this.text}
          <slot />
        </button>
      </Host>
    ) as JSX.Element;
  }
}
