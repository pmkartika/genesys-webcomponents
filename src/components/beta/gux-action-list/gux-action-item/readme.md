# gux-action-item



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description | Type      | Default     |
| ---------- | ---------- | ----------- | --------- | ----------- |
| `disabled` | `disabled` |             | `boolean` | `false`     |
| `text`     | `text`     |             | `string`  | `undefined` |
| `value`    | --         |             | `unknown` | `undefined` |


## Events

| Event   | Description | Type                   |
| ------- | ----------- | ---------------------- |
| `press` |             | `CustomEvent<unknown>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
