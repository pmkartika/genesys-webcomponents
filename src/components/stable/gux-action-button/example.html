<h1>gux-action-button</h1>
<div id="example" class="layout">
  <div class="row">
    <div class="overlap">
      <h2>Primary</h2>
      <gux-action-button text="Primary" accent="primary" is-open="true">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>

      <gux-action-button text="Primary" accent="primary">
        <gux-action-item text="test" disabled></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item disabled><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Secondary</h2>
      <gux-action-button text="Secondary" accent="secondary">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test 3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Tertiary</h2>
      <gux-action-button text="Tertiary" accent="tertiary">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test 3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Disabled</h2>
      <gux-action-button class="disable-toggle" text="Disabled" disabled>
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
  </div>

  <div class="row">
    <div>
      <h2>Primary</h2>
      <gux-action-button text="Primary" accent="primary">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Secondary</h2>
      <gux-action-button text="Secondary" accent="secondary">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test 3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Tertiary</h2>
      <gux-action-button text="Tertiary" accent="tertiary">
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test 3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
    <div>
      <h2>Disabled</h2>
      <gux-action-button class="disable-toggle" text="Disabled" disabled>
        <gux-action-item text="test"></gux-action-item>
        <gux-action-item text="test2"></gux-action-item>
        <gux-action-item text="test3"></gux-action-item>
        <gux-action-list-divider></gux-action-list-divider>
        <gux-action-item><span>I am a span</span></gux-action-item>
      </gux-action-button>
    </div>
  </div>

  <style
    onload="(function () {
    const example = document.getElementById('example');
    example.addEventListener('press', notify);

    const disableToggles = document.getElementsByClassName('disable-toggle');

    Array.from(disableToggles).forEach((disableToggle) => {
      setInterval(() => {
        disableToggle.disabled = !disableToggle.disabled;
      }, 5000);
    });
  })()"
  >
    .layout {
      display: flex;
      flex-direction: column;
      flex-wrap: nowrap;
      align-content: stretch;
      align-items: flex-start;
      justify-content: space-between;
      height: calc(100% - 64px);
    }

    .row {
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      align-content: stretch;
      align-items: flex-start;
      justify-content: space-around;
      width: 100%;
    }

    .overlap gux-action-button {
      margin-bottom: 4px;
    }
  </style>
</div>
