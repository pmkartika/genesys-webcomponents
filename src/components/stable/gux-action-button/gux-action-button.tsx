import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Listen,
  Prop,
  Watch
} from '@stencil/core';

import { trackComponent } from '../../../usage-tracking';
import { OnClickOutside } from '../../../utils/decorator/on-click-outside';

import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import { GuxButtonAccent, GuxButtonType } from '../gux-button/gux-button.types';

import defaultResources from './i18n/en.json';

@Component({
  styleUrl: 'gux-action-button.less',
  tag: 'gux-action-button',
  shadow: true
})
export class GuxActionButton {
  private dropdownButton: HTMLElement;
  private moveFocusDelay: number = 100;
  private i18n: GetI18nValue;

  @Element()
  private root: HTMLElement;
  actionListElement: HTMLGuxActionListElement;

  /**
   * The component button type
   */
  @Prop()
  type: GuxButtonType = 'button';

  /**
   * Triggered when the menu is open
   */
  @Event()
  open: EventEmitter;

  /**
   * Triggered when the menu is close
   */
  @Event()
  close: EventEmitter;

  /**
   * Triggered when the action button is clicked
   */
  @Event()
  actionClick: EventEmitter;

  /**
   * The component text.
   */
  @Prop()
  text: string;

  /**
   * Disables the action button.
   */
  @Prop()
  disabled: boolean = false;

  /**
   * The component accent (secondary or primary).
   */
  @Prop()
  accent: GuxButtonAccent = 'secondary';

  /**
   * It is used to open or not the list.
   */
  @Prop({ mutable: true })
  isOpen: boolean = false;

  @Listen('keydown')
  handleKeydown(event: KeyboardEvent): void {
    const composedPath = event.composedPath();

    switch (event.key) {
      case 'Escape':
        this.isOpen = false;
        if (composedPath.includes(this.actionListElement)) {
          this.dropdownButton.focus();
        }
        break;
      case 'Tab': {
        this.isOpen = false;
        break;
      }
      case 'ArrowDown':
        event.preventDefault();
        if (composedPath.includes(this.dropdownButton)) {
          this.isOpen = true;
          setTimeout(() => {
            void this.actionListElement.setFocusOnFirstItem();
          }, this.moveFocusDelay);
        }
        break;
      case 'Enter':
        if (composedPath.includes(this.dropdownButton)) {
          this.isOpen = true;
          setTimeout(() => {
            void this.actionListElement.setFocusOnFirstItem();
          }, this.moveFocusDelay);
        }
        break;
    }
  }

  @Listen('keyup')
  handleKeyup(event: KeyboardEvent): void {
    const composedPath = event.composedPath();

    switch (event.key) {
      case ' ':
        if (composedPath.includes(this.dropdownButton)) {
          event.preventDefault();
          this.isOpen = true;
          setTimeout(() => {
            void this.actionListElement.setFocusOnFirstItem();
          }, this.moveFocusDelay);
        }
        break;
    }
  }

  @Watch('disabled')
  watchDisabled(disabled: boolean): void {
    if (disabled) {
      this.isOpen = false;
    }
  }

  @Watch('isOpen')
  watchValue(isOpen: boolean): void {
    if (isOpen) {
      this.open.emit();
    } else {
      this.close.emit();
    }
  }

  @OnClickOutside({ triggerEvents: 'click' })
  onClickOutside(event: MouseEvent): void {
    if (event.relatedTarget === null) {
      this.isOpen = false;
    }
  }

  private toggle(): void {
    if (!this.disabled) {
      this.isOpen = !this.isOpen;
    }
  }

  private onActionClick(): void {
    if (!this.disabled) {
      this.isOpen = false;
      this.actionClick.emit();
    }
  }

  async componentWillLoad(): Promise<void> {
    trackComponent(this.root, { variant: this.type });
    this.i18n = await buildI18nForComponent(this.root, defaultResources);
  }

  render(): JSX.Element {
    return (
      <div class="gux-action-button-container">
        <gux-popup-beta expanded={this.isOpen} disabled={this.disabled}>
          <div slot="target" class="gux-action-button-container">
            <gux-button-slot-beta
              class="gux-action-button"
              accent={this.accent}
            >
              <button
                type={this.type}
                disabled={this.disabled}
                onClick={() => this.onActionClick()}
              >
                {this.text}
              </button>
            </gux-button-slot-beta>

            <gux-button-slot-beta
              class="gux-dropdown-button"
              accent={this.accent}
            >
              <button
                type="button"
                disabled={this.disabled}
                ref={el => (this.dropdownButton = el)}
                onClick={() => this.toggle()}
                aria-haspopup="true"
                aria-expanded={this.isOpen.toString()}
                aria-label={this.i18n('actionButtonDropdown', {
                  buttonTitle: this.text
                })}
              >
                <gux-icon decorative icon-name="chevron-small-down"></gux-icon>
              </button>
            </gux-button-slot-beta>
          </div>

          <gux-action-list
            slot="popup"
            ref={el => (this.actionListElement = el)}
          >
            <slot />
          </gux-action-list>
        </gux-popup-beta>
      </div>
    ) as JSX.Element;
  }
}
