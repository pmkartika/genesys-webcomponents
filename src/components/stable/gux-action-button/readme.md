# gux-action-button
An action button. In order to use this element toggle children must be slotted in.

Example usage
```html
<gux-action-button text="Button Text">
  <gux-action-item value="test" text="test"/>
  <gux-action-item value="test2" text="test2"/>
  <gux-action-item value="test3" text="test3"/>
</gux-action-button>
```

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                  | Type                                     | Default       |
| ---------- | ---------- | -------------------------------------------- | ---------------------------------------- | ------------- |
| `accent`   | `accent`   | The component accent (secondary or primary). | `"primary" \| "secondary" \| "tertiary"` | `'secondary'` |
| `disabled` | `disabled` | Disables the action button.                  | `boolean`                                | `false`       |
| `isOpen`   | `is-open`  | It is used to open or not the list.          | `boolean`                                | `false`       |
| `text`     | `text`     | The component text.                          | `string`                                 | `undefined`   |
| `type`     | `type`     | The component button type                    | `"button" \| "reset" \| "submit"`        | `'button'`    |


## Events

| Event         | Description                                 | Type               |
| ------------- | ------------------------------------------- | ------------------ |
| `actionClick` | Triggered when the action button is clicked | `CustomEvent<any>` |
| `close`       | Triggered when the menu is close            | `CustomEvent<any>` |
| `open`        | Triggered when the menu is open             | `CustomEvent<any>` |


## Dependencies

### Depends on

- [gux-popup-beta](../../beta/gux-popup)
- [gux-button-slot-beta](../../beta/gux-button-slot)
- [gux-icon](../gux-icon)
- [gux-action-list](../../beta/gux-action-list)

### Graph
```mermaid
graph TD;
  gux-action-button --> gux-popup-beta
  gux-action-button --> gux-button-slot-beta
  gux-action-button --> gux-icon
  gux-action-button --> gux-action-list
  style gux-action-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
