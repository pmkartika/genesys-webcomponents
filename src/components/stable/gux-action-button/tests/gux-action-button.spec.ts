import { newSpecPage } from '@stencil/core/testing';
import { GuxActionButton } from '../gux-action-button';
import { GuxActionItem } from '../../../beta/gux-action-list/gux-action-item/gux-action-item';

const components = [GuxActionButton, GuxActionItem];
const html = `
<gux-action-button lang="en" text="Primary" accent="primary">
  <gux-action-item text="test"></gux-action-item>
  <gux-action-item text="test2"></gux-action-item>
  <gux-action-item text="test3"></gux-action-item>
  <gux-list-divider></gux-list-divider>
  <gux-action-item><span>I am a span</span></gux-action-item>
</gux-action-button>
`;
const language = 'en';

describe('gux-action-button', () => {
  it('should build', async () => {
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxActionButton);
  });

  it('renders', async () => {
    const page = await newSpecPage({ components, html, language });

    expect(page.root).toMatchSnapshot();
  });

  it('should fire actionClick event if not disabled', async () => {
    const page = await newSpecPage({ components, html, language });
    const actionClickSpy = jest.fn();

    page.win.addEventListener('actionClick', actionClickSpy);

    const element = document.querySelector('gux-action-button');
    const actionButton = element.shadowRoot.querySelector(
      '.gux-action-button > button'
    );

    actionButton.click();

    expect(actionClickSpy).toHaveBeenCalledTimes(1);
  });

  it('should not fire actionClick event if disabled', async () => {
    const disableHtml = `
    <gux-action-button lang="en" text="Primary" accent="primary" disabled>
      <gux-action-item text="test"></gux-action-item>
      <gux-action-item text="test2"></gux-action-item>
      <gux-action-item text="test3"></gux-action-item>
      <gux-list-divider></gux-list-divider>
      <gux-action-item><span>I am a span</span></gux-action-item>
    </gux-action-button>
    `;
    const page = await newSpecPage({ components, html: disableHtml, language });
    const actionClickSpy = jest.fn();

    page.win.addEventListener('actionClick', actionClickSpy);

    const element = document.querySelector('gux-action-button');
    const actionButton = element.shadowRoot.querySelector(
      '.gux-action-button > button'
    );

    actionButton.click();

    expect(actionClickSpy).toHaveBeenCalledTimes(0);
  });

  it('should fire open and close events if not disabled', async () => {
    const page = await newSpecPage({ components, html, language });
    const openSpy = jest.fn();
    const closeSpy = jest.fn();

    page.win.addEventListener('open', openSpy);
    page.win.addEventListener('close', closeSpy);

    const element = document.querySelector('gux-action-button');
    const dropdownButton = element.shadowRoot.querySelector(
      '.gux-dropdown-button > button'
    );

    dropdownButton.click();
    dropdownButton.click();

    expect(openSpy).toHaveBeenCalledTimes(1);
    expect(closeSpy).toHaveBeenCalledTimes(1);
  });

  it('should not fire open event if disabled', async () => {
    const disableHtml = `
    <gux-action-button lang="en" text="Primary" accent="primary" disabled>
      <gux-action-item text="test"></gux-action-item>
      <gux-action-item text="test2"></gux-action-item>
      <gux-action-item text="test3"></gux-action-item>
      <gux-list-divider></gux-list-divider>
      <gux-action-item><span>I am a span</span></gux-action-item>
    </gux-action-button>
    `;
    const page = await newSpecPage({ components, html: disableHtml, language });
    const openSpy = jest.fn();

    page.win.addEventListener('open', openSpy);

    const element = document.querySelector('gux-action-button');
    const dropdownButton = element.shadowRoot.querySelector(
      '.gux-dropdown-button > button'
    );

    dropdownButton.click();

    expect(openSpy).toHaveBeenCalledTimes(0);
  });

  it('should fire press event if action-item not disabled', async () => {
    const page = await newSpecPage({ components, html, language });
    const pressSpy = jest.fn();

    page.win.addEventListener('press', pressSpy);

    const actionItem = document.querySelector('gux-action-item') as HTMLElement;
    actionItem.click();

    expect(pressSpy).toHaveBeenCalledTimes(1);
  });

  it('should not fire press event if action-item disabled', async () => {
    const disableHtml = `
    <gux-action-button lang="en" text="Primary" accent="primary">
      <gux-action-item text="test" disabled></gux-action-item>
      <gux-action-item text="test2"></gux-action-item>
      <gux-action-item text="test3"></gux-action-item>
      <gux-list-divider></gux-list-divider>
      <gux-action-item><span>I am a span</span></gux-action-item>
    </gux-action-button>
    `;
    const page = await newSpecPage({ components, html: disableHtml, language });
    const pressSpy = jest.fn();

    page.win.addEventListener('press', pressSpy);

    const actionItem = document.querySelector('gux-action-item') as HTMLElement;
    actionItem.click();

    expect(pressSpy).toHaveBeenCalledTimes(0);
  });
});
