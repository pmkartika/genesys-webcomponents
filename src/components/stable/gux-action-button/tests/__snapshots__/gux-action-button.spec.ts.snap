// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-action-button renders 1`] = `
<gux-action-button accent="primary" lang="en" text="Primary">
  <mock:shadow-root>
    <div class="gux-action-button-container">
      <gux-popup-beta>
        <div class="gux-action-button-container" slot="target">
          <gux-button-slot-beta accent="primary" class="gux-action-button">
            <button type="button">
              Primary
            </button>
          </gux-button-slot-beta>
          <gux-button-slot-beta accent="primary" class="gux-dropdown-button">
            <button aria-expanded="false" aria-haspopup="true" aria-label="Dropdown for Primary" type="button">
              <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
            </button>
          </gux-button-slot-beta>
        </div>
        <gux-action-list slot="popup">
          <slot></slot>
        </gux-action-list>
      </gux-popup-beta>
    </div>
  </mock:shadow-root>
  <gux-action-item role="listitem" text="test">
    <mock:shadow-root>
      <button class="gux-action-item">
        test
        <slot></slot>
      </button>
    </mock:shadow-root>
  </gux-action-item>
  <gux-action-item role="listitem" text="test2">
    <mock:shadow-root>
      <button class="gux-action-item">
        test2
        <slot></slot>
      </button>
    </mock:shadow-root>
  </gux-action-item>
  <gux-action-item role="listitem" text="test3">
    <mock:shadow-root>
      <button class="gux-action-item">
        test3
        <slot></slot>
      </button>
    </mock:shadow-root>
  </gux-action-item>
  <gux-list-divider></gux-list-divider>
  <gux-action-item role="listitem">
    <mock:shadow-root>
      <button class="gux-action-item">
        <slot></slot>
      </button>
    </mock:shadow-root>
    <span>
      I am a span
    </span>
  </gux-action-item>
</gux-action-button>
`;
